### ISSUE TITLE
Paginate GenerateKeywordsIdeas results

### ISSUE THREAD

**Issue by rahilwazir** [googleads/google-ads-php#341]
**Describe the problem you are trying to solve**

I'm using [`generateKeywordIdeas()`](https://github.com/googleads/google-ads-php/blob/master/src/Google/Ads/GoogleAds/V3/Services/Gapic/KeywordPlanIdeaServiceGapicClient.php#L219) method to fetch all the keywords, similarly like in Google Keyword Planner UI.

Previously I was using [TargetingIdeaService](https://developers.google.com/adwords/api/docs/reference/v201809/TargetingIdeaService), but since I needed the same output as the Google Keyword Planner UI, I was recommended to use `generateKeywordIdeas()` method.

But now I'm unsure how to go about using the pagination method. ~In the [documentation](https://developers.google.com/google-ads/api/docs/keyword-planning/generate-keyword-ideas#generating_ideas) there is a constant `PAGE_SIZE` defined but it is nowhere being used in the example.~

**Describe the solution you would like**

I want to know, how can I paginate the search results and limit them on initial call with `generateKeywordIdeas()`. Just like I used to do with TargetingIdeaService with Adwords API, like so:

```php
use Google\AdsApi\AdWords\v201809\cm\Paging;
....
..
$selector->setPaging( new Paging( 0, $limit ) );
...
```

**Describe alternatives you've considered**

None.

**Comment by fiboknacky** [comment]
Unfortunately, there is no paging support in the `generateKeywordIdeas()` method.
Is your result very large?

**Comment by rahilwazir** [comment]
@fiboknacky To be honest, it depends I can't say any average number of result but large enough in scale if it will be used more than one client.
Right now I'm testing with a specific keyword which yields ~1400 search results.

I know I can paginate after I receive the response but seriously we also want to minimize the number of operations (indirectly the costs) later down the road.

Is it intentional that the support of pagination is not added?

**Comment by fiboknacky** [comment]
> Right now I'm testing with a specific keyword which yields ~1400 search results.

This seems manageable by non-paging requests and could probably be held in memory without issues.

> Is it intentional that the support of pagination is not added?

I cannot answer this question as it's more related to the API design.
If you face with the performance issues with fetching results from this service and need support for the paging mechanism, please feel free to drop feedback in the [Google Ads API forum](https://groups.google.com/forum/?nomobile=true#!forum/adwords-api).

**Comment by rahilwazir** [comment]
> This seems manageable by non-paging requests and could probably be held in memory without issues.

So it appears that it is upto the application to add support for pagination but not provided by the API.

**Comment by fiboknacky** [comment]
> So it appears that it is upto the application to add support for pagination but not provided by the API.

Yes, unfortunately, the current API doesn't provide support for pagination.

**Comment by rahilwazir** [comment]
@fiboknacky In the July update under [Planning](https://developers.google.com/google-ads/api/docs/release-notes#planning) section, it says

> Added paging support to the GenerateKeywordIdeas method.

Can you please update documentation or provide example of how to use the pagination parameters with `generateKeywordIdeas()` method?

**Comment by fiboknacky** [comment]
The paging mechanism should be very similar to that of [`GoogleAdsService::search()`](https://github.com/googleads/google-ads-php/blob/master/src/Google/Ads/GoogleAds/V4/Services/Gapic/GoogleAdsServiceGapicClient.php#L255).

Could you please refer to [this example](https://github.com/googleads/google-ads-php/blob/4d0bb232f6bbc29b5b588929c6862b6f4d2b6b79/examples/AdvancedOperations/FindAndRemoveCriteriaFromSharedSet.php#L122)?
Basically, you just need to pass a page size and then [iterate through your results](https://github.com/googleads/google-ads-php/blob/4d0bb232f6bbc29b5b588929c6862b6f4d2b6b79/examples/AdvancedOperations/FindAndRemoveCriteriaFromSharedSet.php#L126).

**Comment by rahilwazir** [comment]
@fiboknacky Thanks, but for some reason the `pageSize` is not working. It just returns all the results.

```php
$response = $keywordPlanIdeaServiceClient->generateKeywordIdeas(
	new StringValue( [ 'value' => ResourceNames::forLanguageConstant( $languageId ) ] ),
	[
		'customerId'         => $customerId,
		'geoTargetConstants' => $geoTargetConstants,
		'keywordPlanNetwork' => KeywordPlanNetwork::GOOGLE_SEARCH,
		'pageSize'           => 20
	]
);

$results = $response->iterateAllElements();

foreach ( $results as $result ) {
...
```

Total number of `$result` rendered are **119** not **20**. I suppose it is due to `iterateAllElements()` method, it basically ignores the `pageSize`?

The method `$response->getPage()` contains the respected pageSize results i.e. **20**

**Comment by fiboknacky** [comment]
> Total number of $result rendered are 119 not 20. I suppose it is due to iterateAllElements() method, it basically ignores the pageSize?

Yes, sorry that I forgot to mention.
If fetching pages is important to your use case, you would need to use `iteragePages()` instead.
